crypto/rsa/rsa_lib.c

```diff
@@ -906,6 +906,55 @@ int ossl_rsa_get0_all_params(RSA *r, STACK_OF(BIGNUM_const) *primes,
     return 1;
 }
 
+#define safe_BN_num_bits(_k_)  (((_k_) == NULL) ? 0 : BN_num_bits((_k_)))
+int ossl_rsa_check_factors(RSA *r)
+{
+    int valid = 0;
+    int n, d_ok, factors_ok, exps_ok, coeffs_ok, bits, i;
+    STACK_OF(BIGNUM_const) *factors = sk_BIGNUM_const_new_null();
+    STACK_OF(BIGNUM_const) *exps = sk_BIGNUM_const_new_null();
+    STACK_OF(BIGNUM_const) *coeffs = sk_BIGNUM_const_new_null();
+
+    if (factors == NULL || exps == NULL || coeffs == NULL)
+        goto done;
+
+    /*
+     * Simple sanity check for RSA key. All RSA key parameters
+     * must be less-than/equal-to RSA parameter n.
+     */
+    ossl_rsa_get0_all_params(r, factors, exps, coeffs);
+    n = safe_BN_num_bits(RSA_get0_n(r));
+
+    d_ok = (safe_BN_num_bits(RSA_get0_d(r)) <= n);
+
+    exps_ok = 1;
+    for (i = 0; i < sk_BIGNUM_const_num(exps); i++) {
+        bits = safe_BN_num_bits(sk_BIGNUM_const_value(exps, i));
+        exps_ok = ((exps_ok) && (bits <= n));
+    }
+
+    factors_ok = 1;
+    for (i = 0; i < sk_BIGNUM_const_num(factors); i++) {
+        bits = safe_BN_num_bits(sk_BIGNUM_const_value(factors, i));
+        factors_ok = ((factors_ok) && (bits <= n));
+    }
+
+    coeffs_ok = 1;
+    for (i = 0; i < sk_BIGNUM_const_num(coeffs); i++) {
+        bits = safe_BN_num_bits(sk_BIGNUM_const_value(coeffs, i));
+        coeffs_ok = ((coeffs_ok) && (bits <= n));
+    }
+
+    valid = (d_ok && exps_ok && factors_ok && coeffs_ok);
```

Comment:
why not do it simple:
```C
 valid = 0;
 for (...) {
  bits = ...;
   if (bits>n)
     goto done;
}

  valid = 1;
done:
```

Reply:
The inspiration here comes from `BN_mod_exp_mont_consttime()`. Perhaps const time does not matter here. I will change it.